app/controllers/application_controller.rb

```diff
@@ -5,6 +5,159 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, 'yaaaaaaaaaaaaaaaaaaaaas_queeeeeeeeeeen'
   end
 
-end
\ No newline at end of file
+  # Fwitter Homepage
+  get '/' do
+    erb :index
+  end
+
+  ############# TWEETS #####################
+
+  # READ
+  get '/tweets' do #all tweets
+     if is_logged_in? #if person is logged in...
+      @tweets = Tweet.all # get all the tweets
+      erb :'tweets/tweets'
+    else
+      redirect to '/login' #if not, have them log in
+    end
+  end
+
+  # CREATE
+  get '/tweets/new' do #goes to create page/form
+    if is_logged_in?
+      erb :'tweets/create_tweet' #allow them to create a tweet if they are logged in
+    else
+      redirect to '/login'
+    end
+  end
+
+  post '/tweets' do #takes params from the form
+    if params[:content] == '' #no blank tweets allowed. descriptive error message would be nice
+      redirect to 'tweets/new'
+    else
+      user = User.find_by_id(session[:user_id])
+      @tweet = Tweet.create(content: params[:content], user_id: user.id) #make that tweet
+      redirect to "/tweets/#{@tweet.id}" #don't use '' for string interpolation
+    end
+  end
+
+  # READ
+  get '/tweets/:id' do  #get tweets/id should come after get tweets/new
+    if is_logged_in?
+      @tweet = Tweet.find_by_id(params[:id])   #if logged in, find the tweet & show the tweet
+      erb :'tweets/show_tweet'
+    else
+      redirect to '/login'
+    end
+  end
+
+  # UPDATE / EDIT
+    get '/tweets/:id/edit' do  #edit comes after  get tweets/id
+    if is_logged_in?
+      @tweet = Tweet.find_by_id(params[:id])  #find the tweet
+      if @tweet.user_id == session[:user_id] #check that the logged in user is the tweet owner
+       erb :'tweets/edit_tweet'   #then allow them to edit
+      else
+        redirect to '/tweets'  #then take back to user's tweets
+      end
+    else
+      redirect to '/login'
+    end
+  end
+
+  patch '/tweets/:id' do
+    if params[:content] == '' #don't allow them update to a blank tweet
+      redirect to "/tweets/#{params[:id]}/edit"
+    else
+      @tweet = Tweet.find_by_id(params[:id]) #get the tweet
+      @tweet.content = params[:content]  #change the content to params
+      @tweet.save  #save the tweet to the db
+      redirect to "/tweets/#{@tweet.id}" #show the tweet
+    end
+  end
+
+  # DELETE
+  delete '/tweets/:id/delete' do
+    if is_logged_in?
+      @tweet = Tweet.find_by_id(params[:id])  #find that tweet
+      if @tweet.user_id == session[:user_id] #check that the person is the creator
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.